controllers/volumereplicationgroup_controller.go

```diff
@@ -112,24 +111,63 @@ func pvcPredicateFunc() predicate.Funcs {
 
 			log.Info("Update event for PersistentVolumeClaim")
 
-			// TODO: If finalizers change then deep equal of spec fails to catch it, we may want more
-			// conditions here, compare finalizers and also status.state to catch bound PVCs
-			return !reflect.DeepEqual(oldPVC.Spec, newPVC.Spec)
+			return updateEventDecision(oldPVC, newPVC, log)
 		},
 		DeleteFunc: func(e event.DeleteEvent) bool {
-			log := ctrl.Log.WithName("pvcmap").WithName("VolumeReplicationGroup")
-
-			log.Info("delete event for PersistentVolumeClaim")
-
-			// PVC deletes are not of interest, as the added finalizers by VRG would be removed
-			// when VRGs are deleted, which would trigger the delete of the PVC.
+			// No need to reconcile VRG on pvc deletion.
+			// An update event with Nonzero deletion time
+			// would have been received to properly reconcile
+			// VRG.
 			return false
 		},
 	}
 
 	return pvcPredicate
 }
 
+func updateEventDecision(oldPVC *corev1.PersistentVolumeClaim,
+	newPVC *corev1.PersistentVolumeClaim,
+	log logr.Logger) bool {
+	const requeue bool = true
+
+	pvcNamespacedName := types.NamespacedName{Name: newPVC.Name, Namespace: newPVC.Namespace}
+	predicateLog := log.WithValues("pvc", pvcNamespacedName.String())
+	// If finalizers change then deep equal of spec fails to catch it, we may want more
+	// conditions here, compare finalizers and also status.phase to catch bound PVCs
+	if !reflect.DeepEqual(oldPVC.Spec, newPVC.Spec) {
+		predicateLog.Info("Spec changed. Requeing")
+
+		return requeue
+	}
+
+	if oldPVC.Status.Phase != corev1.ClaimBound && newPVC.Status.Phase == corev1.ClaimBound {
+		predicateLog.Info("PersistentVolumeClaim bound. ", "oldPVC Phase", oldPVC.Status.Phase,
+			"newPVC phase", newPVC.Status.Phase)
+
+		return requeue
+	}
+
+	// First check whether the pvc is being deleted or not. Then check for
+	// finalizer's presence. Because, if the pvc is being deleted and as part
+	// of handling it, finalizer might have been deleted in a previous reconcile
+	// loop. So, no need to add finalizer if a pvc is being deleted.
```

Comment:
Can we fold the conditions below to trigger it based on [this](https://github.com/RamenDR/ramen/blob/65ff8515076501398f41cc9b4a6bc2d3ac4ba38e/controllers/volumereplicationgroup_controller.go#L597-L602) check instead? (and make that check a function to enable re-use)

Reply:
ACK

Reply:
We should call `isPVCDeletedAndNotProtected` here to ensure that PVC is protected before requeueing?

Currently if a PVC is deleted then VRG is reconciled due to the condition here. We want to reconcile if PVC is deleted and was originally protected, and skip if not.

Also, I was not able to get the comment:
```
        // First check whether the pvc is being deleted or not. Then check for
	// finalizer's presence. Because, if the pvc is being deleted and as part
	// of handling it, finalizer might have been deleted in a previous reconcile
	// loop. So, no need to add finalizer if a pvc is being deleted.
 ```

We are not adding a finalizer here, so that part in the comment did not parse.

We remove the finalizer last, so once it is removed and PVC is deleted we are no longer interested in the PVC event, hence both go hand in hand.

The corner case of VR delete failing for a PVC once it is deleted may not be triggered by a PVC delete, and needs separate handling.